test/common/http/http3_status_tracker_test.cc

```diff
@@ -0,0 +1,181 @@
+#include "common/http/http3_status_tracker.h"
+
+#include "test/mocks/common.h"
+#include "test/mocks/event/mocks.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using Envoy::Event::MockTimer;
+using testing::NiceMock;
+using testing::Return;
+using testing::StrictMock;
+
+namespace Envoy {
+namespace Http {
+
+namespace {
+class Http3StatusTrackerTest : public testing::Test {
+public:
+  Http3StatusTrackerTest()
+      : timer_(new StrictMock<MockTimer>(&dispatcher_)), tracker_(dispatcher_) {}
+
+  NiceMock<Event::MockDispatcher> dispatcher_;
+  StrictMock<MockTimer>* timer_; // Owned by tracker_;
+  Http3StatusTracker tracker_;
+};
+
+TEST_F(Http3StatusTrackerTest, Initialized) {
+  EXPECT_FALSE(tracker_.isHttp3Broken());
+  EXPECT_FALSE(tracker_.isHttp3Confirmed());
+}
+
+TEST_F(Http3StatusTrackerTest, MarkBroken) {
+  EXPECT_CALL(*timer_, enabled()).WillOnce(Return(false));
+  EXPECT_CALL(*timer_, enableTimer(std::chrono::milliseconds(5 * 60 * 1000), nullptr));
+  tracker_.markHttp3Broken();
+  EXPECT_TRUE(tracker_.isHttp3Broken());
+  EXPECT_FALSE(tracker_.isHttp3Confirmed());
+}
+
+TEST_F(Http3StatusTrackerTest, MarkBrokenRepeatedly) {
+  EXPECT_CALL(*timer_, enabled()).WillOnce(Return(false));
+  EXPECT_CALL(*timer_, enableTimer(std::chrono::milliseconds(5 * 60 * 1000), nullptr));
+  tracker_.markHttp3Broken();
+  EXPECT_TRUE(tracker_.isHttp3Broken());
+  EXPECT_FALSE(tracker_.isHttp3Confirmed());
+
+  EXPECT_CALL(*timer_, enabled()).WillOnce(Return(true));
+  tracker_.markHttp3Broken();
+  EXPECT_TRUE(tracker_.isHttp3Broken());
+  EXPECT_FALSE(tracker_.isHttp3Confirmed());
+}
+
+TEST_F(Http3StatusTrackerTest, MarkBrokenThenExpires) {
+  EXPECT_CALL(*timer_, enabled()).WillOnce(Return(false));
+  EXPECT_CALL(*timer_, enableTimer(std::chrono::milliseconds(5 * 60 * 1000), nullptr));
+  tracker_.markHttp3Broken();
+
+  timer_->invokeCallback();
+  EXPECT_FALSE(tracker_.isHttp3Broken());
+  EXPECT_FALSE(tracker_.isHttp3Confirmed());
+}
+
+TEST_F(Http3StatusTrackerTest, MarkBrokenWithBackoff) {
+  // markBroken will only be called when the time is not enabled.
```

Comment:
From reading the test it seems like we're the ones calling `markHttp3Broken`, what does this comment refer to?

Reply:
Ah. I can remove this comment if it doesn't make sense. I wrote the comment because the MockTimer API confused me. I expected that expect() would return true if enableTimer() had been called but disableTimer() or invokeCallback() had not been. In other words, I didn't expect to need to mock out this method. So the comment was saying that when markBroken() is called in this tests, the timer will not have been enabled. Does that make sense? Would you recommend I rephrase or remove the comment?

Reply:
Oh I see, I would try to include something that indicates that we're talking about invariants for this is used by prod code.  My initial read was that this was trying to explain what we should expect to see in this test, so I expected to see some `EXPECT_CALL(.., markBroken())` kind of expectations

Reply:
Makes sense. That being said, I took several stabs at trying to write something up and each time ended up with a bit of an essay that didn't really seem to add any readability to the test, so I've just nuked the comment. (Really the problem was my lack of understanding of how MockTimer worked and so this is probably not the best place to address that :>)